File: bindings/csharp/WickraImpact/Impact.cs
using System.Runtime.InteropServices;
using System.Text;

namespace Wickra.Impact;

/// <summary>
/// A market-impact backtest driven by JSON commands, built from a spec,
/// over the Wickra C ABI. Construct one from a spec JSON, drive it with command
/// JSON (<c>set_spec</c>, <c>run</c>, <c>version</c>) and read
/// back the response JSON — the same protocol as the CLI and every other binding.
/// </summary>
public sealed class Impact : IDisposable
{
    private readonly ImpactHandle _handle;

    /// <summary>Build a backtest handle from a spec JSON (<c>"{}"</c> defers to a later <c>set_spec</c>).</summary>
    /// <exception cref="ArgumentException">The spec was not a valid backtest spec.</exception>
    public Impact(string specJson)
    {
        IntPtr ptr = Native.wickra_impact_new(Utf8(specJson));
        if (ptr == IntPtr.Zero)
        {
            throw new ArgumentException("wickra-impact: invalid spec", nameof(specJson));
        }
        _handle = new ImpactHandle(ptr);
    }

    /// <summary>Apply a command JSON and return the response JSON.</summary>
    /// <remarks>
    /// Uses the C ABI's length-out protocol: a first call learns the length, then
    /// the response is read into a caller-owned buffer. Domain errors (a bad
    /// command, an unknown command) come back in-band as <c>{"ok":false,...}</c>
    /// JSON, not as an exception.
    /// </remarks>
    /// <exception cref="InvalidOperationException">A required argument was unusable or a panic was caught.</exception>
    public string Command(string cmdJson)
    {
        ObjectDisposedException.ThrowIf(_handle.IsInvalid, this);

        byte[] cmd = Utf8(cmdJson);
        IntPtr h = _handle.DangerousGetHandle();
        int n = Native.wickra_impact_command(h, cmd, null, 0);
        if (n < 0)
        {
            throw new InvalidOperationException($"wickra-impact: command failed (code {n})");
        }
        var buf = new byte[n + 1];
        Native.wickra_impact_command(h, cmd, buf, (nuint)buf.Length);
        return Encoding.UTF8.GetString(buf, 0, n);
    }

    /// <summary>The library version.</summary>
    public static string Version() =>
        Marshal.PtrToStringUTF8(Native.wickra_impact_version()) ?? string.Empty;

    /// <summary>Free the native backtest handle.</summary>
    public void Dispose() => _handle.Dispose();

    /// <summary>Encode a string as NUL-terminated UTF-8 for the C ABI.</summary>
    private static byte[] Utf8(string s)
    {
        int len = Encoding.UTF8.GetByteCount(s);
        var buf = new byte[len + 1];
        Encoding.UTF8.GetBytes(s, 0, s.Length, buf, 0);
        return buf;
    }
}

/// <summary>A safe handle owning a native backtest pointer.</summary>
internal sealed class ImpactHandle : SafeHandle
{
    public ImpactHandle(IntPtr handle)
        : base(IntPtr.Zero, ownsHandle: true) => SetHandle(handle);

    public override bool IsInvalid => handle == IntPtr.Zero;

    protected override bool ReleaseHandle()
    {
        Native.wickra_impact_free(handle);
        return true;
    }
}

File: bindings/java/src/main/java/org/wickra/impact/Impact.java
package org.wickra.impact;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;

/**
 * A market-impact backtest driven by JSON commands, built from a spec,
 * over the Wickra C ABI (FFM/Panama). Construct one from a spec JSON, drive it
 * with command JSON ({@code set_spec}, {@code run}, {@code version}) and read
 * back the response JSON — the same protocol as the CLI and every other binding.
 */
public final class Impact implements AutoCloseable {
    private MemorySegment handle;

    /**
     * Build a backtest handle from a spec JSON ({@code "{}"} defers configuration
     * to a later {@code set_spec} command).
     *
     * @throws IllegalArgumentException if the spec is not a valid backtest spec
     */
    public Impact(String specJson) {
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment spec = arena.allocateFrom(specJson);
            MemorySegment created = (MemorySegment) Native.NEW.invokeExact(spec);
            if (created.address() == 0) {
                throw new IllegalArgumentException("wickra-impact: invalid spec");
            }
            this.handle = created;
        } catch (RuntimeException | Error e) {
            throw e;
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /**
     * Apply a command JSON and return the response JSON. Uses the C ABI's
     * length-out protocol: a first call learns the length, then the response is
     * read into a caller-owned buffer. Domain errors (a bad command, an unknown
     * command) come back in-band as {@code {"ok":false,...}} JSON, not as an
     * exception.
     */
    public String command(String cmdJson) {
        if (handle == null) {
            throw new IllegalStateException("impact is closed");
        }
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment cmd = arena.allocateFrom(cmdJson);
            int len = (int) Native.COMMAND.invokeExact(handle, cmd, MemorySegment.NULL, 0L);
            if (len < 0) {
                throw new IllegalStateException("wickra-impact: command failed (code " + len + ")");
            }
            MemorySegment buf = arena.allocate(len + 1L);
            int written = (int) Native.COMMAND.invokeExact(handle, cmd, buf, (long) (len + 1));
            return buf.getString(0);
        } catch (RuntimeException | Error e) {
            throw e;
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /** The library version. */
    public static String version() {
        try {
            MemorySegment ptr = (MemorySegment) Native.VERSION.invokeExact();
            return ptr.reinterpret(Long.MAX_VALUE).getString(0);
        } catch (Throwable t) {
            throw new RuntimeException(t);
        }
    }

    /** Free the native backtest handle. */
    @Override
    public void close() {
        if (handle != null) {
            try {
                Native.FREE.invokeExact(handle);
            } catch (Throwable t) {
                throw new RuntimeException(t);
            }
            handle = null;
        }
    }
}
